src/cascadia/TerminalSettingsEditor/ActionsViewModel.cpp

```diff
@@ -18,383 +24,1194 @@ using namespace winrt::Windows::UI::Xaml::Data;
+        _AvailableActionsAndNamesMap{ availableShortcutActionsAndNames }
+    {
+    }
 
-    KeyBindingViewModel::KeyBindingViewModel(const Control::KeyChord& keys, const hstring& actionName, const IObservableVector<hstring>& availableActions) :
-        _CurrentKeys{ keys },
-        _KeyChordText{ KeyChordSerialization::ToString(keys) },
-        _CurrentAction{ actionName },
-        _ProposedAction{ box_value(actionName) },
-        _AvailableActions{ availableActions }
+    void CommandViewModel::Initialize()
     {
+        std::vector<Editor::KeyChordViewModel> keyChordVMs;
+        for (const auto keys : _keyChordList)
+        {
+            auto kcVM{ make<KeyChordViewModel>(keys) };
+            _RegisterKeyChordVMEvents(kcVM);
+            keyChordVMs.push_back(kcVM);
+        }
+        _KeyChordViewModelList = single_threaded_observable_vector(std::move(keyChordVMs));
+
+        std::vector<hstring> shortcutActions;
+        for (const auto [action, name] : _AvailableActionsAndNamesMap)
+        {
+            shortcutActions.emplace_back(name);
+            _NameToActionMap.emplace(name, action);
+        }
+        std::sort(shortcutActions.begin(), shortcutActions.end());
+        _AvailableShortcutActions = single_threaded_observable_vector(std::move(shortcutActions));
+
+        const auto shortcutActionString = _AvailableActionsAndNamesMap.Lookup(_command.ActionAndArgs().Action());
+        ProposedShortcutAction(winrt::box_value(shortcutActionString));
+        _CreateAndInitializeActionArgsVMHelper();
+
         // Add a property changed handler to our own property changed event.
-        // This propagates changes from the settings model to anybody listening to our
-        //  unique view model members.
+        // This allows us to create a new ActionArgsVM when the shortcut action changes
         PropertyChanged([this](auto&&, const PropertyChangedEventArgs& args) {
             const auto viewModelProperty{ args.PropertyName() };
-            if (viewModelProperty == L"CurrentKeys")
+            if (viewModelProperty == L"ProposedShortcutAction")
             {
-                _KeyChordText = KeyChordSerialization::ToString(_CurrentKeys);
-                _NotifyChanges(L"KeyChordText");
+                const auto actionString = unbox_value<hstring>(ProposedShortcutAction());
+                const auto actionEnum = _NameToActionMap.at(actionString);
+                const auto emptyArgs = CascadiaSettings::GetEmptyArgsForAction(actionEnum);
+                // todo: probably need some better default values for empty args
+                // eg. for sendInput, where "input" is a required argument, "input" gets set to an empty string which does not satisfy the requirement
+                // i.e. if the user hits "save" immediately after switching to sendInput as the action (without adding something to the input field), they'll get an error
+                // there are some other cases as well

```

Comment:
Bare TODO. Good thinking. Definitely worth mentioning it as a follow-up.

We have a few places throughout the settings UI that could benefit from validation. Like, all the regular expression stuff I added to the new tab menu's profile matcher, file/folder paths, etc. It's a larger issue that would definitely be nice to figure out.